src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/LockFactory.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Razor;
+
+internal class LockFactory
+{
+    // Specify recursion is supported, since an item with an upgradeable lock can still
+    // get another read lock on the same thread
+    protected readonly ReaderWriterLockSlim _lock = new(LockRecursionPolicy.SupportsRecursion);
+    private static readonly TimeSpan s_timeout = TimeSpan.FromMinutes(2);
+
+    public IDisposable GetReadLock() => new ReadOnlyLock(_lock);
+    public IDisposable GetWriteLock() => new WriteOnlyLock(_lock);
+    public UpgradeAbleReadLock GetUpgradeAbleReadLock() => new UpgradeAbleReadLock(_lock);
+
+    public void EnsureNoWriteLock()
+    {
+        if (_lock.IsWriteLockHeld)
+        {
+            throw new InvalidOperationException("Expected no write lock to be held");
+        }
+    }
+
+    private class ReadOnlyLock : IDisposable
```

Comment:
Is it possible to turn these disposable classes into ref structs to avoid allocation every time we want to lock?